java/src/main/java/cucumber/runtime/java/ObjectFactoryLoader.java

```diff
@@ -32,18 +36,16 @@ private ObjectFactoryLoader() {
     public static ObjectFactory loadObjectFactory(ClassFinder classFinder, String objectFactoryClassName) {
         ObjectFactory objectFactory;
         try {
-            Reflections reflections = new Reflections(classFinder);
-
-            if (objectFactoryClassName != null) {
-                Class<ObjectFactory> objectFactoryClass = (Class<ObjectFactory>) classFinder.loadClass(objectFactoryClassName);
-                objectFactory = reflections.newInstance(new Class[0], new Object[0], objectFactoryClass);
+            final Reflections reflections = new Reflections(classFinder);
+            final ServiceLoader<io.cucumber.core.backend.ObjectFactory> loader = ServiceLoader.load(io.cucumber.core.backend.ObjectFactory.class);
+            if (objectFactoryClassName == null) {
+                objectFactory = new ObjectFactoryAdapter(loadSingleObjectFactory(reflections, loader));
```

Comment:
Adapters should only be applied as needed. Can you push this down to the point it's needed? Same for the one below.

Reply:
Was my first idea, but than I've not found at which place?
1. not necessary for:
1.1 old Interfaces-impl. loaded by ServiceLoader
1.2 old interfaces-impl. loaded by Reflection
2. necessary for:
2.1. only new interface-impl. loaded by ServiceLoader
2.2. only new interface-impl. loaded by Reflection, and yes with current impl. is it possible
This means every finder can return a Factory which need this Adapter, and if you look at the impl. the overhead of it is minimal.

Reply:
It's important to keep related bits of functionality together. Makes it easier to understand the code and once ObjectFactory no longer extends io.cucumber.core.backend.ObjectFactory it won't be possible either anyway.